include/net/can.h

```diff
@@ -0,0 +1,186 @@
+ * +++   +---+   +--+----+   +-------+   +--+----+   +--/    \_/               \
+ * | |        \ /     |   \ /         \ /     |   \ /  /                       |
+ * | |         X      |    X    CAN    X      |    X  |         Internet       |
+ * | |        / \     |   / \         / \     |   / \  \                       /
+ * +++   +---+   +----+--+   +-------+   +----+--+   +--+                     /
+ *  |    |                                               +-------------------+
+ *  +----+
+ */
+struct net_can_api {
+	/**
+	 * The net_if_api must be placed in first position in this
+	 * struct so that we are compatible with network interface API.
+	 */
+	struct net_if_api iface_api;
+
+	/** Send a single CAN frame */
+	int (*send)(struct device *dev, const struct zcan_frame *frame,
+		    can_tx_callback_t cb, void *cb_arg, s32_t timeout);
+	/** Attach a filter with it's callback */
+	int (*attach_filter)(struct device *dev, can_rx_callback_t cb,
+			     void *cb_arg, const struct zcan_filter *filter);
+	/** Detach a filter */
+	void (*detach_filter)(struct device *dev, int filter_id);
+	/** Enable or disable the reception of frames for net CAN */
+	int (*enable)(struct device *dev, bool enable);
+};
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * Context for canbus net device.
+ */
+struct canbus_net_ctx {
+	int dad_filter_id;
+	struct k_work dad_work;
+	struct net_if *iface;
+	u16_t ll_addr;
+};
+
+/**
+ * Canbus link layer addresses have a length of 14 bit for source and destination.
+ * Both together are 28 bit to fit a CAN extended identifier with 29 bit length.
+ */
+struct net_can_lladdr {
+	u16_t addr : 14;
+};
+
+/*
+ * STmin is split in two ranges:
+ *   0-127: 0ms-127ms
+ * 128-240: Reserved
+ * 241-249: 100us-900us (multiples of 100us)
+ * 250-   : Reserved
+ */
+struct can_net_fc_opts {
+	u8_t bs;     /** Block size. Number of CF PDUs before next CF is sent */
+	u8_t stmin;  /** Minimum separation time. Min time between frames */
+};
+
+/**
+ * Context for a transmission of messages that didn't fit in a single frame.
+ * This messages Start with a FF (First Frame) that is in case of unicast
+ * acknowledged by a FC (Frame Control). After that N CF (Consecutive frames)
+ * carry the rest of the message.
+ */
+struct can_net_isotp_tx_ctx {
+	/** Frame Control options received from FC frame */
+	struct can_net_fc_opts opts;
+	/** State of the transmission */
+	u8_t state;
+	/** Actual block number that is transmitted. Counts from BS to 0 */
+	u8_t act_block_nr;
+	/** Number of WAIT frames received*/
+	u8_t wft;
+	/** Sequence number that is added to CF*/
+	u8_t sn : 4;
+	/** Transmission is multicast */
+	u8_t is_mcast : 1;
+	/** Remaining data to transbit in bytes */
+	size_t rem_len;
+	/** Timeout for TX Timeout and separation time*/
+	struct _timeout timeout;
+	/** Pkt containing the data to transmit */
+	struct net_pkt *pkt;
+};
+
+
+struct can_net_isotp_rx_ctx {
+	/** Remaining data to receive. Goes from message length to zero */
```

Comment:
same here

Reply:
4 bit holes Can't be avoided i guess ;-)

Reply:
Yep :) But typically we can save bytes for each struct by carefully placing the fields in proper order. Not sure if this an issue here, it just looked there might be some optimization possibilities.

Reply:
I changed the order. pahole does not show XXX any more